tokio-postgres/tests/test/replication.rs

```diff
@@ -0,0 +1,149 @@
+        .unwrap();
+
+    let slot = "test_logical_slot";
+
+    let query = format!(
+        r#"CREATE_REPLICATION_SLOT {:?} TEMPORARY LOGICAL "pgoutput""#,
+        slot
+    );
+    let slot_query = client.simple_query(&query).await.unwrap();
+    let lsn = if let Row(row) = &slot_query[0] {
+        row.get("consistent_point").unwrap()
+    } else {
+        panic!("unexpeced query message");
+    };
+
+    // issue a query that will appear in the slot's stream since it happened after its creation
+    client
+        .simple_query("INSERT INTO test_logical_replication VALUES (42)")
+        .await
+        .unwrap();
+
+    let options = r#"("proto_version" '1', "publication_names" 'test_pub')"#;
+    let query = format!(
+        r#"START_REPLICATION SLOT {:?} LOGICAL {} {}"#,
+        slot, lsn, options
+    );
+    let copy_stream = client
+        .copy_both_simple::<bytes::Bytes>(&query)
+        .await
+        .unwrap();
+
+    let stream = LogicalReplicationStream::new(copy_stream);
+    tokio::pin!(stream);
+
+    // verify that we can observe the transaction in the replication stream
+    let begin = loop {
+        match stream.next().await {
+            Some(Ok(XLogData(body))) => {
+                if let Begin(begin) = body.into_data() {
+                    break begin;
+                }
+            }
+            Some(Ok(_)) => (),
+            Some(Err(_)) => panic!("unexpected replication stream error"),
+            None => panic!("unexpected replication stream end"),
+        }
+    };
+
+    let insert = loop {
+        match stream.next().await {
+            Some(Ok(XLogData(body))) => {
+                if let Insert(insert) = body.into_data() {
+                    break insert;
+                }
+            }
+            Some(Ok(_)) => (),
+            Some(Err(_)) => panic!("unexpected replication stream error"),
+            None => panic!("unexpected replication stream end"),
+        }
+    };
+
+    let commit = loop {
+        match stream.next().await {
+            Some(Ok(XLogData(body))) => {
+                if let Commit(commit) = body.into_data() {
+                    break commit;
+                }
+            }
+            Some(Ok(_)) => (),
+            Some(Err(_)) => panic!("unexpected replication stream error"),
+            None => panic!("unexpected replication stream end"),
+        }
+    };
+
+    assert_eq!(begin.final_lsn(), commit.commit_lsn());
+    assert_eq!(insert.rel_id(), rel_id);
+
+    let tuple_data = insert.tuple().tuple_data();
+    assert_eq!(tuple_data.len(), 1);
+    assert!(matches!(tuple_data[0], TupleData::Text(_)));
+    if let TupleData::Text(data) = &tuple_data[0] {
+        assert_eq!(data, &b"42"[..]);
+    }
+
+    // Send a standby status update and require a keep alive response
+    let lsn: PgLsn = lsn.parse().unwrap();
+    let epoch = UNIX_EPOCH + Duration::from_secs(946_684_800);
+    let ts = epoch.elapsed().unwrap().as_micros() as i64;
```

Comment:
What is being computed here? What is the significance of `946_684_800`?

Reply:
All stamps in PG are since internal Epoch which is set to 2000-01-01+00.
You can find this mentioned in the protocol description page: https://www.postgresql.org/docs/13/protocol-replication.html
`The client's system clock at the time of transmission, as microseconds since midnight on 2000-01-01.`


The constant `946_684_800` is a unixtime stamp of the PG's Epoch.

Reply:
added a comment to clarify this

Reply:
Thanks for the clarification.

Perhaps this code should be inside a public `pg_time_now()` function?

Reply:
Actually, something like this would be even better. 

```rust
struct PgTimestamp(u64);
impl From<SystemTime> for PgTimeStamp {
    // ...
}
```

That makes `standby_status_update` a little more foolproof.

Reply:
I wrote a first draft of `PgTimestamp` [here](https://github.com/petrosagg/rust-postgres/pull/1/commits/a0d067b66447951d1276a53fb09886539c3fa094).

Reply:
The commit "use SystemTime for timestamps at API boundary" resolves this question.